docs/source/generics.rst

```diff
@@ -0,0 +1,984 @@
+  can be harder to reason about.
+
+* :py:data:`~typing.Callable` is an example of type that behaves contravariant
+  in types of arguments. That is, ``Callable[[Shape], int]`` is a subtype of
+  ``Callable[[Triangle], int]``, despite ``Shape`` being a supertype of
+  ``Triangle``. To understand this, consider:
+
+  .. code-block:: python
+
+    def cost_of_paint_required(
+        triangle: Triangle,
+        area_calculator: Callable[[Triangle], float]
+    ) -> float:
+        return area_calculator(triangle) * DOLLAR_PER_SQ_FT
+
+    # This straightforwardly works
+    def area_of_triangle(triangle: Triangle) -> float: ...
+    cost_of_paint_required(triangle, area_of_triangle)  # OK
+
+    # But this works as well!
+    def area_of_any_shape(shape: Shape) -> float: ...
+    cost_of_paint_required(triangle, area_of_any_shape)  # OK
+
+  ``cost_of_paint_required`` needs a callable that can calculate the area of a
+  triangle. If we give it a callable that can calculate the area of an
+  arbitrary shape (not just triangles), everything still works.
+
+* :py:class:`~typing.List` is an invariant generic type. Naively, one would think
+  that it is covariant, like :py:class:`~typing.Sequence` above, but consider this code:
+
+  .. code-block:: python
+
+     class Circle(Shape):
+         # The rotate method is only defined on Circle, not on Shape
+         def rotate(self): ...
+
+     def add_one(things: list[Shape]) -> None:
+         things.append(Shape())
+
+     my_circles: list[Circle] = []
+     add_one(my_circles)     # This may appear safe, but...
+     my_circles[-1].rotate()  # ...this will fail, since my_circles[0] is now a Shape, not a Circle
+
+  Another example of invariant type is :py:class:`~typing.Dict`. Most mutable containers
+  are invariant.
+
+By default, all user-defined generics are invariant.
+To declare a given generic class as covariant or contravariant use
+type variables defined with special keyword arguments ``covariant`` or
+``contravariant``. For example:
+
+.. code-block:: python
+
+   from typing import Generic, TypeVar
+
+   T_co = TypeVar('T_co', covariant=True)
+
+   class Box(Generic[T_co]):  # this type is declared covariant
+       def __init__(self, content: T_co) -> None:
+           self._content = content
+
+       def get_content(self) -> T_co:
+           return self._content
+
+   def look_into(box: Box[Animal]): ...
+
+   my_box = Box(Cat())
+   look_into(my_box)  # OK, but would be an error if Box was invariant in T
+
+.. _type-variable-upper-bound:
+
+Type variables with upper bounds
+********************************
+
+By default, a type variable can be replaced with any type. This means that
+you can't do very much with an object of type ``T`` safely -- you don't
+know anything about it!
+
+It's therefore often useful to be able to limit the types that a type
+variable can take on, for instance, by restricting it to values that are
+subtypes of a specific type.
```

Comment:
Nit: "consistent subtypes", since the bound can include `Any`.

Reply:
I actually think it's okay to be a little loose with the terminology here, since this is documentation aimed at beginners. The "consistent" qualifier doesn't mean much unless you've been brushing up on your type theory, and a beginner _probably_ won't have been :)

Reply:
Thanks Alex, I agree that new typing users probably will not care too much about this detail. However, I expect them to be equally unfamiliar with both subtyping and consistent subtyping. So, we will not make this paragraph any harder to follow by using the right term.